// node-irc/index.js
module.exports = function() {
	var net = require('net');
	return {
		Client: function(host, port, options) {
			var client = null;
			var modules = [];
			var moduleData = [];
			options = options || {};
			options.nick = options.nick || 'Moka_Ur';
			var ircClient = {
				connect: function() {
					client = net.connect({
						port: port,
						host: host
					}, function() {
						console.log('Connected to ' + host + ':' + port);
						client.write('NICK ' + options.nick + '\r\n');
						client.write('USER ' + options.nick + ' 8 * :Felhő Úr\r\n');
					});
					var oldWrite = client.write;
					client.write = function(data) {
						ircClient.log('tcp', 'Sending: ' + data);
						oldWrite.call(client, data);
					}
					client.on('data', function(data) {
						data = data.toString();
						//console.log(data);
						var i = 0;
						var next = function() {
								if(modules[i]) {
									modules[i++](data, next);
								}
							}
						next();
					});
				},
				write: function(data) {
					client.write(data);
				},
				use: function(module, opt) {
					if(typeof module == 'string') {
						moduleData.push({
							name: module,
							opt: opt
						});
						console.log('Loading module '+module);
						module = require(module)(ircClient, opt);
					}
					if(typeof module == 'function') {
						modules.push(module);
					}
				},
				reload: function(){
					for(var i in require.cache){
						delete require.cache[i];
					}
					modules=[];
					var oldModules=moduleData.slice(0);
					moduleData=[];
					for(var i in oldModules){
						var m=oldModules[i];
						ircClient.use(m.name,m.opt);
					}
				}
			};
			return ircClient;
		}
	}
}();
